exporter/exporterhelper/wal.go

```diff
@@ -0,0 +1,312 @@
+	"time"
+
+	"github.com/jaegertracing/jaeger/pkg/queue"
+	"go.uber.org/atomic"
+	"go.uber.org/zap"
+
+	"go.opentelemetry.io/collector/extension/storage"
+)
+
+// WALQueue holds the WAL-backed queue
+type WALQueue struct {
+	logger     *zap.Logger
+	numWorkers int
+	retryDelay time.Duration
+	stopWG     sync.WaitGroup
+	stopped    *atomic.Uint32
+	storage    walStorage
+}
+
+type walStorage interface {
+	put(ctx context.Context, req request) error
+	get(ctx context.Context) (request, error)
+	size() int
+	stop()
+}
+
+// walContiguousStorage provides a WAL implementation backed by file storage extension
+//
+// Write index describes the position at which next item is going to be stored
+// Read index describes which item needs to be read next
+// When Write index = Read index, no elements are in the queue
+//
+//   ┌───────WAL-backed queue──────┐
+//   │                             │
+//   │     ┌───┐     ┌───┐ ┌───┐   │
+//   │ n+1 │ n │ ... │ 4 │ │ 3 │   │
+//   │     └───┘     └───┘ └─x─┘   │
+//   │                       x     │
+//   └───────────────────────x─────┘
+//      ▲              ▲     x
+//      │              │     xxx deleted
+//      │              │
+//    write          read
+//    index          index
+//
+type walContiguousStorage struct {
+	logger        *zap.Logger
+	client        storage.Client
+	mu            sync.Mutex
+	id            string
+	readIndex     uint64
+	writeIndex    uint64
+	readIndexKey  string
+	writeIndexKey string
+	unmarshaler   requestUnmarshaler
+}
+
+const (
+	zapItemKey         = "itemKey"
+	zapWalIDKey        = "walID"
+	itemKeyTemplate    = "%s-it-%d"
+	readIndexTemplate  = "%s-ri"
+	writeIndexTemplate = "%s-wi"
+	defaultRetryDelay  = 1 * time.Second
+)
+
+var (
+	errNotFound = errors.New("not found")
+)
+
+func newWALQueue(ctx context.Context, id string, logger *zap.Logger, client storage.Client, unmarshaler requestUnmarshaler) *WALQueue {
+	return &WALQueue{
+		logger:     logger,
+		stopped:    atomic.NewUint32(0),
+		retryDelay: defaultRetryDelay,
+		storage:    newWALContiguousStorage(ctx, id, logger, client, unmarshaler),
+	}
+}
+
+// StartConsumers starts the given number of consumers which will be consuming items
+func (wq *WALQueue) StartConsumers(num int, callback func(item interface{})) {
+	wq.numWorkers = num
+	var startWG sync.WaitGroup
+
+	factory := func() queue.Consumer {
+		return queue.ConsumerFunc(callback)
+	}
+
+	for i := 0; i < wq.numWorkers; i++ {
+		wq.stopWG.Add(1)
+		startWG.Add(1)
+		go func() {
+			startWG.Done()
+			defer wq.stopWG.Done()
+			consumer := factory()
+
+			for {
+				if wq.stopped.Load() != 0 {
+					return
+				}
+				req, err := wq.storage.get(context.Background())
+				if err == errNotFound || req == nil {
+					time.Sleep(wq.retryDelay)
+				} else {
+					consumer.Consume(req)
+				}
+			}
+		}()
+	}
+	startWG.Wait()
+}
+
+// Produce adds an item to the queue and returns true if it was accepted
+func (wq *WALQueue) Produce(item interface{}) bool {
```

Comment:
Should this be changed to be more type safe? 
```suggestion
func (wq *WALQueue) Produce(item request) bool {
```

Reply:
Actually, I kept it like that to match the BoundedQueue interface